File: lib/python/opentoken/src/main/opentoken/keyexchange/key_pair_manager.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import logging
from pathlib import Path
from typing import Tuple

from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.backends import default_backend

from opentoken.keyexchange.key_exchange_exception import KeyExchangeException


logger = logging.getLogger(__name__)


class KeyPairManager:
    """
    Manages ECDH key pair lifecycle for OpenToken.

    Handles generation, loading, and saving of elliptic curve key pairs
    used for Diffie-Hellman key exchange. Uses the P-256 (secp256r1) curve
    for broad compatibility and security.
    """

    DEFAULT_KEY_DIR = os.path.join(os.path.expanduser("~"), ".opentoken")
    DEFAULT_PRIVATE_KEY_FILENAME = "keypair.pem"
    DEFAULT_PUBLIC_KEY_FILENAME = "public_key.pem"
    EC_CURVE = ec.SECP256R1()

    def __init__(self, key_directory: str = None):
        """
        Creates a KeyPairManager.

        Args:
            key_directory: The directory to store keys. Defaults to ~/.opentoken.
        """
        self.key_directory = key_directory or self.DEFAULT_KEY_DIR

    def get_or_create_key_pair(self) -> Tuple[ec.EllipticCurvePrivateKey, ec.EllipticCurvePublicKey]:
        """
        Loads or generates a key pair.

        If a key pair exists at the default location, it is loaded.
        Otherwise, a new key pair is generated and saved.

        Returns:
            A tuple of (private_key, public_key).

        Raises:
            KeyExchangeException: If key loading or generation fails.
        """
        private_key_path = os.path.join(self.key_directory, self.DEFAULT_PRIVATE_KEY_FILENAME)

        if os.path.exists(private_key_path):
            logger.info(f"Loading existing key pair from {private_key_path}")
            return self.load_key_pair(private_key_path)
        else:
            logger.info("No existing key pair found. Generating new key pair.")
            return self.generate_and_save_key_pair()

    def generate_key_pair(self) -> Tuple[ec.EllipticCurvePrivateKey, ec.EllipticCurvePublicKey]:
        """
        Generates a new ECDH key pair using the P-256 curve.

        Returns:
            A tuple of (private_key, public_key).

        Raises:
            KeyExchangeException: If key generation fails.
        """
        try:
            private_key = ec.generate_private_key(self.EC_CURVE, default_backend())
            public_key = private_key.public_key()

            logger.debug("Generated new ECDH key pair using curve SECP256R1")
            return private_key, public_key
        except Exception as e:
            raise KeyExchangeException("Failed to generate ECDH key pair", e)

    def generate_and_save_key_pair(self) -> Tuple[ec.EllipticCurvePrivateKey, ec.EllipticCurvePublicKey]:
        """
        Generates a new key pair and saves it to the default location.

        Returns:
            A tuple of (private_key, public_key).

        Raises:
            KeyExchangeException: If generation or saving fails.
        """
        private_key, public_key = self.generate_key_pair()
        self.save_key_pair(private_key, public_key)
        return private_key, public_key

    def save_key_pair(self, private_key: ec.EllipticCurvePrivateKey, 
                     public_key: ec.EllipticCurvePublicKey) -> None:
        """
        Saves a key pair to the configured directory.

        The private key is saved with restrictive permissions (0o600 on Unix-like systems).

        Args:
            private_key: The private key to save.
            public_key: The public key to save.

        Raises:
            KeyExchangeException: If saving fails.
        """
        try:
            # Create directory if it doesn't exist
            os.makedirs(self.key_directory, exist_ok=True)
            logger.debug(f"Ensured key directory exists: {self.key_directory}")

            # Save private key
            private_key_path = os.path.join(self.key_directory, self.DEFAULT_PRIVATE_KEY_FILENAME)
            self.save_private_key(private_key, private_key_path)

            # Save public key
            public_key_path = os.path.join(self.key_directory, self.DEFAULT_PUBLIC_KEY_FILENAME)
            self.save_public_key(public_key, public_key_path)

            logger.info(f"Saved key pair to {self.key_directory}")
        except Exception as e:
            raise KeyExchangeException("Failed to save key pair", e)

    def save_private_key(self, private_key: ec.EllipticCurvePrivateKey, file_path: str) -> None:
        """
        Saves a private key to a file in PEM format with restrictive permissions.

        Args:
            private_key: The private key to save.
            file_path: The path to save the key.

        Raises:
            KeyExchangeException: If saving fails.
        """
        try:
            pem = private_key.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption()
            )

            with open(file_path, 'wb') as f:
                f.write(pem)

            # Set restrictive permissions (Unix-like systems only)
            try:
                os.chmod(file_path, 0o600)
                logger.debug(f"Set restrictive permissions (0600) on {file_path}")
            except Exception as e:
                logger.warning(f"Failed to set restrictive permissions on {file_path}: {e}")

            logger.debug(f"Saved private key to {file_path}")
        except Exception as e:
            raise KeyExchangeException(f"Failed to save private key to {file_path}", e)

    def save_public_key(self, public_key: ec.EllipticCurvePublicKey, file_path: str) -> None:
        """
        Saves a public key to a file in PEM format.

        Args:
            public_key: The public key to save.
            file_path: The path to save the key.

        Raises:
            KeyExchangeException: If saving fails.
        """
        try:
            pem = public_key.public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo
            )

            with open(file_path, 'wb') as f:
                f.write(pem)

            logger.debug(f"Saved public key to {file_path}")
        except Exception as e:
            raise KeyExchangeException(f"Failed to save public key to {file_path}", e)

    def load_key_pair(self, private_key_path: str) -> Tuple[ec.EllipticCurvePrivateKey, ec.EllipticCurvePublicKey]:
        """
        Loads a key pair from the specified private key file path.

        This method loads the private key from the specified path and attempts to load
        the corresponding public key from the same directory (using the standard public key filename).

        Args:
            private_key_path: The path to the private key file.

        Returns:
            A tuple of (private_key, public_key).

        Raises:
            KeyExchangeException: If loading fails.
        """
        try:
            # Load private key
            with open(private_key_path, 'rb') as f:
                private_key = serialization.load_pem_private_key(
                    f.read(),
                    password=None,
                    backend=default_backend()
                )

            # Determine public key path (same directory, standard filename)
            private_path = Path(private_key_path)
            public_key_path = private_path.parent / self.DEFAULT_PUBLIC_KEY_FILENAME

            # Load public key
            if public_key_path.exists():
                with open(public_key_path, 'rb') as f:
                    public_key = serialization.load_pem_public_key(
                        f.read(),
                        backend=default_backend()
                    )
            else:
                raise KeyExchangeException(
                    f"Public key file not found at expected location: {public_key_path}"
                )

            return private_key, public_key
        except Exception as e:
            if isinstance(e, KeyExchangeException):
                raise
            raise KeyExchangeException(f"Failed to load key pair from {private_key_path}", e)

    def get_key_directory(self) -> str:
        """
        Returns the key directory path.

        Returns:
            The key directory.
        """
        return self.key_directory

File: lib/java/opentoken/src/main/java/com/truveta/opentoken/keyexchange/KeyPairManager.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.keyexchange;

import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.attribute.PosixFilePermission;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.NoSuchAlgorithmException;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.spec.ECGenParameterSpec;
import java.security.spec.InvalidKeySpecException;
import java.util.HashSet;
import java.util.Set;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Manages ECDH key pair lifecycle for OpenToken.
 * <p>
 * Handles generation, loading, and saving of elliptic curve key pairs
 * used for Diffie-Hellman key exchange. Uses the P-256 (secp256r1) curve
 * for broad compatibility and security.
 */
public class KeyPairManager {
    private static final Logger logger = LoggerFactory.getLogger(KeyPairManager.class);
    
    /**
     * Default directory for storing key pairs.
     */
    public static final String DEFAULT_KEY_DIR = System.getProperty("user.home") + "/.opentoken";
    
    /**
     * Default filename for private key.
     */
    public static final String DEFAULT_PRIVATE_KEY_FILENAME = "keypair.pem";
    
    /**
     * Default filename for public key.
     */
    public static final String DEFAULT_PUBLIC_KEY_FILENAME = "public_key.pem";
    
    /**
     * Elliptic curve algorithm name.
     */
    public static final String EC_ALGORITHM = "EC";
    
    /**
     * Standard curve for ECDH (P-256 / secp256r1).
     */
    public static final String EC_CURVE = "secp256r1";
    
    private final String keyDirectory;
    
    /**
     * Creates a KeyPairManager with the default key directory.
     */
    public KeyPairManager() {
        this(DEFAULT_KEY_DIR);
    }
    
    /**
     * Creates a KeyPairManager with a custom key directory.
     *
     * @param keyDirectory the directory to store keys
     */
    public KeyPairManager(String keyDirectory) {
        this.keyDirectory = keyDirectory;
    }
    
    /**
     * Loads or generates a key pair.
     * <p>
     * If a key pair exists at the default location, it is loaded.
     * Otherwise, a new key pair is generated and saved.
     *
     * @return the key pair (loaded or newly generated)
     * @throws KeyExchangeException if key loading or generation fails
     */
    public KeyPair getOrCreateKeyPair() throws KeyExchangeException {
        Path privateKeyPath = Paths.get(keyDirectory, DEFAULT_PRIVATE_KEY_FILENAME);
        
        if (Files.exists(privateKeyPath)) {
            logger.info("Loading existing key pair from {}", privateKeyPath);
            return loadKeyPair(privateKeyPath.toString());
        } else {
            logger.info("No existing key pair found. Generating new key pair.");
            return generateAndSaveKeyPair();
        }
    }
    
    /**
     * Generates a new ECDH key pair using the P-256 curve.
     *
     * @return the newly generated key pair
     * @throws KeyExchangeException if key generation fails
     */
    public KeyPair generateKeyPair() throws KeyExchangeException {
        try {
            KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance(EC_ALGORITHM);
            ECGenParameterSpec ecSpec = new ECGenParameterSpec(EC_CURVE);
            keyPairGenerator.initialize(ecSpec, new SecureRandom());
            
            KeyPair keyPair = keyPairGenerator.generateKeyPair();
            logger.debug("Generated new ECDH key pair using curve {}", EC_CURVE);
            
            return keyPair;
        } catch (NoSuchAlgorithmException | java.security.InvalidAlgorithmParameterException e) {
            throw new KeyExchangeException("Failed to generate ECDH key pair", e);
        }
    }
    
    /**
     * Generates a new key pair and saves it to the default location.
     *
     * @return the newly generated key pair
     * @throws KeyExchangeException if generation or saving fails
     */
    public KeyPair generateAndSaveKeyPair() throws KeyExchangeException {
        KeyPair keyPair = generateKeyPair();
        saveKeyPair(keyPair);
        return keyPair;
    }
    
    /**
     * Saves a key pair to the configured directory.
     * <p>
     * The private key is saved with restrictive permissions (0600 on Unix-like systems).
     *
     * @param keyPair the key pair to save
     * @throws KeyExchangeException if saving fails
     */
    public void saveKeyPair(KeyPair keyPair) throws KeyExchangeException {
        try {
            // Create directory if it doesn't exist
            Path keyDirPath = Paths.get(keyDirectory);
            if (!Files.exists(keyDirPath)) {
                Files.createDirectories(keyDirPath);
                logger.debug("Created key directory: {}", keyDirectory);
            }
            
            // Save private key
            Path privateKeyPath = Paths.get(keyDirectory, DEFAULT_PRIVATE_KEY_FILENAME);
            savePrivateKey(keyPair.getPrivate(), privateKeyPath.toString());
            
            // Save public key
            Path publicKeyPath = Paths.get(keyDirectory, DEFAULT_PUBLIC_KEY_FILENAME);
            savePublicKey(keyPair.getPublic(), publicKeyPath.toString());
            
            logger.info("Saved key pair to {}", keyDirectory);
        } catch (IOException e) {
            throw new KeyExchangeException("Failed to save key pair", e);
        }
    }
    
    /**
     * Saves a private key to a file in PEM format with restrictive permissions.
     *
     * @param privateKey the private key to save
     * @param filePath the path to save the key
     * @throws KeyExchangeException if saving fails
     */
    public void savePrivateKey(PrivateKey privateKey, String filePath) throws KeyExchangeException {
        try {
            String pemContent = PemUtils.encodeToPem(privateKey, "PRIVATE KEY");
            
            // Write to file
            try (FileOutputStream fos = new FileOutputStream(filePath)) {
                fos.write(pemContent.getBytes());
            }
            
            // Set restrictive permissions (Unix-like systems only)
            setRestrictivePermissions(filePath);
            
            logger.debug("Saved private key to {}", filePath);
        } catch (IOException e) {
            throw new KeyExchangeException("Failed to save private key", e);
        }
    }
    
    /**
     * Saves a public key to a file in PEM format.
     *
     * @param publicKey the public key to save
     * @param filePath the path to save the key
     * @throws KeyExchangeException if saving fails
     */
    public void savePublicKey(PublicKey publicKey, String filePath) throws KeyExchangeException {
        try {
            String pemContent = PemUtils.encodeToPem(publicKey, "PUBLIC KEY");
            
            try (FileOutputStream fos = new FileOutputStream(filePath)) {
                fos.write(pemContent.getBytes());
            }
            
            logger.debug("Saved public key to {}", filePath);
        } catch (IOException e) {
            throw new KeyExchangeException("Failed to save public key", e);
        }
    }
    
    /**
     * Loads a key pair from the specified private key file path.
     * <p>
     * This method loads the private key from the specified path and attempts to load
     * the corresponding public key from the same directory (using the standard public key filename).
     *
     * @param privateKeyPath the path to the private key file
     * @return the loaded key pair
     * @throws KeyExchangeException if loading fails
     */
    public KeyPair loadKeyPair(String privateKeyPath) throws KeyExchangeException {
        try {
            PrivateKey privateKey = PemUtils.loadPrivateKey(privateKeyPath);
            
            // Determine public key path (same directory, standard filename)
            Path privatePath = Paths.get(privateKeyPath);
            Path publicKeyPath = privatePath.getParent().resolve(DEFAULT_PUBLIC_KEY_FILENAME);
            
            PublicKey publicKey;
            if (Files.exists(publicKeyPath)) {
                publicKey = PemUtils.loadPublicKey(publicKeyPath.toString());
            } else {
                throw new KeyExchangeException(
                    "Public key file not found at expected location: " + publicKeyPath
                );
            }
            
            return new KeyPair(publicKey, privateKey);
        } catch (IOException | InvalidKeySpecException | NoSuchAlgorithmException e) {
            throw new KeyExchangeException("Failed to load key pair from " + privateKeyPath, e);
        }
    }
    
    /**
     * Sets restrictive file permissions (0600) on Unix-like systems.
     * <p>
     * On Windows, this method does nothing as the permission model is different.
     *
     * @param filePath the path to the file
     */
    private void setRestrictivePermissions(String filePath) {
        try {
            Path path = Paths.get(filePath);
            
            // Check if we're on a POSIX-compliant system
            if (path.getFileSystem().supportedFileAttributeViews().contains("posix")) {
                Set<PosixFilePermission> permissions = new HashSet<>();
                permissions.add(PosixFilePermission.OWNER_READ);
                permissions.add(PosixFilePermission.OWNER_WRITE);
                
                Files.setPosixFilePermissions(path, permissions);
                logger.debug("Set restrictive permissions (0600) on {}", filePath);
            } else {
                logger.debug("POSIX permissions not supported on this system, skipping permission setting");
            }
        } catch (IOException e) {
            logger.warn("Failed to set restrictive permissions on {}: {}", filePath, e.getMessage());
        }
    }
    
    /**
     * Returns the key directory path.
     *
     * @return the key directory
     */
    public String getKeyDirectory() {
        return keyDirectory;
    }
}
